apps/arweave/src/ar_pool.erl

```diff
@@ -0,0 +1,764 @@
+-export([init/1, handle_cast/2, handle_call/3, handle_info/2, terminate/2]).
+
+-include_lib("arweave/include/ar_config.hrl").
+-include_lib("arweave/include/ar_consensus.hrl").
+-include_lib("arweave/include/ar_mining.hrl").
+-include_lib("arweave/include/ar_pool.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+-record(state, {
+	%% The most recent keys come first.
+	session_keys = [],
+	%% Key => [{Output, StepNumber, PartitionUpperBound, Seed, Diff}, ...]
+	jobs_by_session_key = maps:new(),
+	request_pid_by_ref = maps:new()
+}).
+
+%%%===================================================================
+%%% Public interface.
+%%%===================================================================
+
+%% @doc Start the server.
+start_link() ->
+	gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+%% @doc Return true if we are a pool client.
+is_client() ->
+	{ok, Config} = application:get_env(arweave, config),
+	Config#config.pool_client == true.
+
+%% @doc Return a list of up to two most recently cached VDF session keys.
+get_current_sessions() ->
+	gen_server:call(?MODULE, get_current_sessions, infinity).
+
+%% @doc Return a set of the most recent cached jobs.
+get_jobs(PrevOutput) ->
+	gen_server:call(?MODULE, {get_jobs, PrevOutput}, infinity).
+
+%% @doc Return the most recent cached VDF output. Return an empty binary if the
+%% cache is empty.
+get_latest_output() ->
+	gen_server:call(?MODULE, get_latest_output, infinity).
+
+%% @doc Cache the given jobs.
+cache_jobs(Jobs) ->
+	gen_server:cast(?MODULE, {cache_jobs, Jobs}).
+
+%% @doc Validate the given (partial) solution. If the solution is eligible for
+%% producing a block, produce and publish a block.
+process_partial_solution(Solution) ->
+	gen_server:call(?MODULE, {process_partial_solution, Solution}, infinity).
+
+%% @doc Send the partial solution to the pool.
+post_partial_solution(Solution) ->
+	gen_server:cast(?MODULE, {post_partial_solution, Solution}).
+
+%%%===================================================================
+%%% Generic server callbacks.
+%%%===================================================================
+
+init([]) ->
+	process_flag(trap_exit, true),
+	ok = ar_events:subscribe(solution),
+	{ok, #state{}}.
+
+handle_call(get_current_sessions, _From, State) ->
+	{reply, lists:sublist(State#state.session_keys, 2), State};
+
+handle_call({get_jobs, PrevOutput}, _From, State) ->
+	SessionKeys = State#state.session_keys,
+	JobCache = State#state.jobs_by_session_key,
+	{reply, get_jobs(PrevOutput, SessionKeys, JobCache), State};
+
+handle_call(get_latest_output, _From, State) ->
+	case State#state.session_keys of
+		[] ->
+			{reply, <<>>, State};
+		[Key | _] ->
+			{O, _SN, _U, _S, _Diff} = hd(maps:get(Key, State#state.jobs_by_session_key)),
+			{reply, O, State}
+	end;
+
+handle_call({process_partial_solution, Solution}, From, State) ->
+	#state{ request_pid_by_ref = Map } = State,
+	Ref = make_ref(),
+	case process_partial_solution(Solution, Ref) of
```

Comment:
The `process_partial_solution` series of checks mirrors `prepare_solution` and `post_solution` checks in `ar_mining_server` in that it performs a series of validations and then ends by sending a `found_solution` event. Any way we can align the meaning to make that similarity/symmetry more explicit? E.g. rename `process_partial_solutino` to `prepare` or `post` - or rename the functions in `ar_mining_server` to `process`

Also `ar_mining_server` uses `atom` matching to differentiate different steps in the process (e.g. `prepare_solution(poa2, ....)` vs. here adds onto the name of the fucntion (e.g. `process_partial_solution_poa2(...)`

Can we align those two conventions? either both use `atom` or both use function name? I think it will also make reviewing/updating/understanding the code easier in the future.

Reply:
Hm, I see it also mirrors and duplicates a lot of checks in `ar_block_pre_validator`... any chance we can deduplicate those two code paths? If you want to punt all these refactoring changes to me, I can take a look

Reply:
I'd say they have structural similarities but are semantically different so different names and some code duplication are justified - `prepare_solution` builds one whereas `process_partial_solution` processes an external solution.

I prefer different functions to atom matching because the linter can catch mistakes (by highlighting unused or undefined functions). I had a temptation to refactor prepare_solution but was discouraged by the way `prepare_solution(poa1` and `prepare_solution(poa2` are intertwined [here](https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_mining_server.erl#L567).

Reply:
Got it. Is `process_partial_solution` basically validating the partial solution? And in that way perhaps is more semantically similar to the code in `ar_block_pre_validator` and `ar_node_utils`? If so I can add a note to to https://github.com/ArweaveTeam/arweave-dev/issues/354 to look into deduplicating common code there.

Ultimately I'm selfishly trying to reduce work for myself in the future - I've spent a lot of time tracing through the different validation paths, sometimes picking the wrong set of validation code to look at, etc... If I can take the 4 different validation paths we have now (`ar_pool`, `ar_block_pre_validator`, `ar_node_utils`, `ar_mining_server`) and unify at least parts of them, it'll reduce the time I spend in the future debugging and reading code.

Also: I added a note to the refactoring issue [here](https://github.com/ArweaveTeam/arweave-dev/issues/354) to also remove the atom matching from `prepare_solution`